Source/libDolphin/libDolphin.cpp

```diff
@@ -0,0 +1,431 @@
+// Copyright 2008 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+// This is the library to communicate with Dolphin.
+
+#include "Stdafx.h"
+#include "libDolphin.h"
+#include <Windows.h>
+
+struct GCPadStatus m_current_pad;
+void(*m_callback) ();
+
+HANDLE h_pipe;
+
+bool Dolphin::Init() {
+	h_pipe = CreateFile(
+		LPCWSTR(PIPE_NAME),
+		GENERIC_READ | GENERIC_WRITE,
+		0,
+		NULL,
+		OPEN_EXISTING,
+		0,
+		NULL
+	);
+
+	char buffer[2];
+	DWORD bytes;
+	buffer[0] = 0x0B;
+
+	WriteFile(
+		h_pipe,
+		buffer,
+		2,
+		&bytes,
+		NULL
+	);
+
+	ReadFile(
+		h_pipe,
+		buffer,
+		2,
+		&bytes,
+		NULL
+	);
+
+	if ((bytes != 2) || (buffer[0] != 0x1B)) {
+		return false;
+	}
+
+	return true;
+};
+
+void Dolphin::SetFrameCallback(void(*callback)()) {
+	m_callback = callback;
+};
+
+void RunFrameCallback(void(*callback) ()) {
+	(*callback)();
+}
+
+// Potential return values
+// -1: Didn't set the frame callback.
+// -2: Invalid response from server.
+int Dolphin::StartLoop() {
+	if (m_callback == NULL) {
+		return -1;
+	}
+
+	while (true) {
+		m_current_pad = EMPTY_CONTROL;
+
+		RunFrameCallback(m_callback);
+		u64 *input;
+
+		input = InterpretInput(m_current_pad);
+
+		char *buffer;
+
+		buffer[0] = 0x09;
+
+		u64 *ubuffer = (u64*) buffer + 1;
+
+		ubuffer[0] = input[0];
+		ubuffer[1] = input[1];
+
+		char rbuffer[2];
+		DWORD bytes;
+
+		WriteFile(
+			h_pipe,
+			buffer,
+			18,
+			&bytes,
+			NULL
+		);
+
+		ReadFile(
+			h_pipe,
+			rbuffer,
+			2,
+			&bytes,
+			NULL
+		);
+
+		if ((bytes != 2) || (rbuffer[0] != 0x19)) {
+			return -2;
+		}
+
+		char buffer2[2];
+
+		buffer2[0] = 0x0A;
+
+		WriteFile(
+			h_pipe,
+			buffer2,
+			2,
+			&bytes,
+			NULL
+		);
+
+		ReadFile(
+			h_pipe,                // handle to pipe 
+			buffer,             // buffer to receive data 
+			2,     // size of buffer 
+			&bytes,             // number of bytes read 
+			NULL
+		);
+
+		if ((bytes != 2) || (buffer[0] != 0x1A)) {
+			return -2;
+		}
+	}
+}
+
+u64* Dolphin::InterpretInput(struct GCPadStatus pad) {
```

Comment:
struct is not necessary in C++, you can simply declare it as GCPadStatus, and it should be passed by const reference.